src/Objectivity.AutoFixture.XUnit2.Core/SpecimenBuilders/RandomFixedValuesGenerator.cs

```diff
@@ -0,0 +1,50 @@
+﻿namespace Objectivity.AutoFixture.XUnit2.Core.SpecimenBuilders
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Concurrent;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Linq;
+
+    using global::AutoFixture.Kernel;
+
+    using Objectivity.AutoFixture.XUnit2.Core.Common;
+    using Objectivity.AutoFixture.XUnit2.Core.Requests;
+
+    internal sealed class RandomFixedValuesGenerator : ISpecimenBuilder
+    {
+        private readonly ConcurrentDictionary<FixedValuesRequest, IEnumerator> enumerators = new();
+
+        public object Create(object request, ISpecimenContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (request.NotNull(nameof(request)) is FixedValuesRequest fixedValuesRequest)
+            {
+                return this.CreateValue(fixedValuesRequest);
+            }
+
+            return new NoSpecimen();
+        }
+
+        [SuppressMessage("Security", "CA5394:Do not use insecure randomness", Justification = "It is good enough for collection randomization.")]
+        private static IEnumerator CreateEnumerable(FixedValuesRequest request)
+        {
+            var random = new Random();
+            var values = request.Values.OrderBy((_) => random.Next()).ToArray();
+
+            return new RoundRobinEnumerable<object>(values).GetEnumerator();
+        }
+
+        private object CreateValue(FixedValuesRequest request)
+        {
+            var generator = this.enumerators.GetOrAdd(request, CreateEnumerable);
+            generator.MoveNext();
```

Comment:
This way you may end up returning the same value multiple times, if a method is called more than the number of enum values.
I also don't like the sorting, can't we simply pick a random element out of an array of enum values?

Reply:
I prefer a solution that seeks to exhaust a set of specified values first. Such design is inlined with AutoFixture [EnumGenerator](https://github.com/AutoFixture/AutoFixture/blob/master/Src/AutoFixture/EnumGenerator.cs).